src/topology/alexandroff.lean

```diff
@@ -0,0 +1,319 @@
+* The connectedness of `alexandroff X` for a noncompact, preconnected `X`
+* `alexandroff X` is `T₁` for a T₁ space `X`
+* `alexandroff X` is Hausdorff if `X` is locally compact and Hausdorff
+-/
+
+noncomputable theory
+open filter set
+open_locale classical topological_space filter
+
+section basic
+
+/-- The Alexandroff extension of an arbitrary topological space `X` -/
+@[nolint unused_arguments]
+def alexandroff (X : Type*) [topological_space X] := option X
+
+variables {X : Type*} [topological_space X]
+
+/-- The embedding of `X` to its Alexandroff extension -/
+def of : X → alexandroff X := some
+
+/-- The range of the embedding -/
+def range_of (X : Type*) [topological_space X] : set (alexandroff X) := of '' (univ : set X)
+
+lemma of_apply {x : X} : of x = some x := rfl
+
+lemma of_injective : function.injective (@of X _) :=
+option.some_injective X
+
+/-- The point at infinity -/
+def infty : alexandroff X := none
+localized "notation `∞` := infty" in alexandroff
+
+namespace alexandroff
+
+instance : has_coe_t X (alexandroff X) := ⟨of⟩
+
+instance : inhabited(alexandroff X) := ⟨∞⟩
+
+@[norm_cast]
+lemma coe_eq_coe {x y : X} : (x : alexandroff X) = y ↔ x = y :=
+of_injective.eq_iff
+
+@[simp] lemma coe_ne_infty (x : X) : (x : alexandroff X) ≠ ∞  .
+@[simp] lemma infty_ne_coe (x : X) : ∞ ≠ (x : alexandroff X) .
+@[simp] lemma of_eq_coe {x : X} : (of x : alexandroff X) = x := rfl
+
+/-- Recursor for `alexandroff` using the preferred forms `∞` and `↑x`. -/
+@[elab_as_eliminator]
+def rec_infty_coe (C : alexandroff X → Sort*) (h₁ : C infty) (h₂ : Π (x : X), C x) :
+  Π (z : alexandroff X), C z :=
+option.rec h₁ h₂
+
+lemma ne_infty_iff_exists {x : alexandroff X} :
+  x ≠ ∞ ↔ ∃ (y : X), x = y :=
+by { induction x using alexandroff.rec_infty_coe; simp }
+
+@[simp] lemma coe_mem_range_of (x : X) : (x : alexandroff X) ∈ (range_of X) :=
+by simp [range_of]
+
+lemma union_infty_eq_univ : (range_of X ∪ {∞}) = univ :=
+begin
+  refine le_antisymm (subset_univ _) (λ x hx, _),
+  induction x using alexandroff.rec_infty_coe; simp
+end
+
+@[simp] lemma infty_not_mem_range_of : ∞ ∉ range_of X :=
+by simp [range_of]
+
+@[simp] lemma not_mem_range_of_iff (x : alexandroff X) :
+  x ∉ range_of X ↔ x = ∞ :=
+by { induction x using alexandroff.rec_infty_coe; simp [infty_not_mem_range_of] }
+
+attribute [nolint simp_nf] not_mem_range_of_iff
+
+lemma infty_not_mem_image_of {s : set X} : ∞ ∉ of '' s :=
+not_mem_subset (image_subset _ $ subset_univ _) infty_not_mem_range_of
+
+lemma inter_infty_eq_empty : (range_of X) ∩ {∞} = ∅ :=
+by { ext x, induction x using alexandroff.rec_infty_coe; simp }
+
+lemma of_preimage_infty : (of⁻¹' {∞} : set X) = ∅ :=
+by { ext, simp }
+
+end alexandroff
+
+end basic
+
+open alexandroff
+open_locale alexandroff
+
+section topology
+
+variables {X : Type*} [topological_space X]
+
+instance : topological_space (alexandroff X) :=
+{ is_open := λ s, if ∞ ∈ s then is_compact (of⁻¹' s)ᶜ ∧ is_open (of⁻¹' s)
+    else is_open (of⁻¹' s),
```

Comment:
Possibly easier to work with as
```suggestion
{ is_open := λ s, (∞ ∈ s → is_compact (of ⁻¹' s)ᶜ) ∧ is_open (of ⁻¹' s),
```
obviously the proofs below would have to change.

Reply:
Would this force us to use more `by_cases` in the proof? With the former definition, we can use `split_ifs` quite conveniently.

Reply:
I expect it would make things simpler because you no longer have to prove `is_open (of⁻¹' s)` twice

Reply:
I still think the `if then else` definition is better. In most cases, `is_open (of⁻¹' s)` can be proved with two super short `exact` statements because the key argument of the proof can be done on `if... then... else...`, and we no longer need to write `by_cases h : ∞ ∈ s` and specialize the definition to get `is_compact (of ⁻¹' s)ᶜ`.

Reply:
Have a look at my suggested proofs - they all _remove_ `by_cases h : ∞ ∈ s` rather than adding it.

Reply:
Wow they are super elegant.